tests/31sync/03joined.pl

```diff
@@ -124,3 +124,65 @@
          Future->done(1)
       })
    };
+
+test "Newly joined room has correct timeline in incremental sync",
+   requires => [qw( first_api_client can_sync )],
+
+   check => sub {
+      my ( $http ) = @_;
+
+      my ( $user_a, $filter_id_a, $user_b, $filter_id_b, $room_id, $next_b );
+
+      my $filter = {
+         room => {
+            timeline => { types => [ "m.room.message" ], limit => 10 },
+            state    => { types => [] },
+         }
+      };
+
+      Future->needs_all(
+         matrix_register_user_with_filter( $http, $filter ),
+         matrix_register_user_with_filter( $http, $filter ),
+      )->then( sub {
+         ( $user_a, $filter_id_a, $user_b, $filter_id_b ) = @_;
+
+         matrix_create_room( $user_a );
+      })->then( sub {
+         ( $room_id ) = @_;
+
+         Future->needs_all( map {
+            matrix_send_room_text_message( $user_a, $room_id, body => "test" );
+         } 0 .. 3 );
+      })->then( sub {
+         matrix_sync( $user_b, filter => $filter_id_b );
+      })->then( sub {
+         my ( $body ) = @_;
+
+         $next_b = $body->{next_batch};
+
+         Future->needs_all( map {
+            matrix_send_room_text_message( $user_a, $room_id, body => "test" );
+         } 0 .. 3 );
+      })->then( sub {
+         matrix_join_room( $user_b, $room_id );
+      })->then( sub {
+         matrix_sync( $user_b, filter => $filter_id_b, since => $next_b );
+      })->then( sub {
+         my ( $body ) = @_;
+         my $room = $body->{rooms}{joined}{$room_id};
+         my $timeline = $room->{timeline};
+
+         if( @{ $timeline->{events} } == 6 ) {
+            # We could assert that the timeline wasn't limited in this case
```

Comment:
Let's in fact assert that $timeline->{limited} == JSON::false, and maybe also assert that all six events are m.room.message events (i.e. verify that we're actually filtering out the state event from the join)